langlib/langlib-test/src/test/resources/test-src/stringlib_test.bal

```diff
@@ -31,24 +31,39 @@ function testSubString() returns [string,string, string] {
     return [str.substring(6, 9), str.substring(6), strings:substring(str,6)];
 }
 
-function testIterator() returns string[] {
+function testIterator() {
     string str = "Foo Bar";
-
-    object {
-         public isolated function next() returns record {| string value; |}?;
-    } itr = str.iterator();
-
-    string[] chars = [];
-    int i = 0;
-    record {| string value; |}|() elem = itr.next();
-
-    while (elem is record {| string value; |}) {
-        chars[i] = elem.value;
-        elem = itr.next();
-        i += 1;
+    string:Char[] expected = ["F", "o", "o", " ", "B", "a", "r"];
+    int i;
+
+    foreach int test in 1 ..< 3 {
+        i = 0;
+        match test{
+            1 => {
+                object {
+                    public isolated function next() returns record {| string:Char value; |}?;
+                    } itr = str.iterator();
+                record {| string:Char value; |}|() elem = itr.next();
+                while (elem is record {| string:Char value; |}) {
+                    assertEquals(expected[i], elem.value);
+                    elem = itr.next();
+                    i += 1;
+                }
+            }
+
+            2 => {
+                object {
+                    public isolated function next() returns record {| string value; |}?;
+                    } itr = str.iterator();
+                record {| string value; |}|() elem = itr.next();
+                while (elem is record {| string:Char value; |}) {
+                    assertEquals(expected[i], elem.value);
+                    elem = itr.next();
+                    i+= 1;
+                }
+            }
+        }
     }
```

Comment:
Why do we need a foreach statement here. Can't we just do 

```ballerina
    object {
        public isolated function next() returns record {| string:Char value; |}?;
    } itr1 = str.iterator();
    record {| string:Char value; |}|() elem1 = itr1.next();
    while (elem1 is record {| string:Char value; |}) {
        assertEquals(expected[i], elem1.value);
        elem1 = itr1.next();
        i += 1;
    }
    assertEquals(7, i);

    i = 0;
    object {
        public isolated function next() returns record {| string value; |}?;
    } itr2 = str.iterator();
    record {| string value; |}|() elem2 = itr2.next();
    while (elem2 is record {| string:Char value; |}) {
        assertEquals(expected[i], elem2.value);
        elem2 = itr2.next();
        i += 1;
    }
    assertEquals(7, i);
```

We should also assert `i` to verify that the expected number of iterations happen. If there's eventually a change that breaks the type of the `value` field in the returned record to `string`, the current test wouldn't catch it since it'll simply not go into the while loops.